File: BOJ/python2/1874.py
n = int(raw_input())
S = []
i = 1
for _ in xrange (n):
    print S
    x = int(raw_input())
    if x not in S:
        while True:
            S.append(i)
            print '+'
            i += 1
            if x in S:
                S.pop()
                print '-'
                break
    else:
        if S[-1] != x:
            print 'NO'
            break
        else:
            S.pop()
            print '-'

File: BOJ/cpp/1874.cpp
#include <stdio.h>
#include <stack>
using namespace std;

int main(){
  int n;
  scanf("%d",&n);
  char result[3*n+100000];
  int arr[n+3];
  for (int i=1;i<=n;i++){
    scanf("%d",&arr[i]);
  }
  arr[n+1] = -1;
  stack <int> s;
  int i = 1, status = 1, resultIndex = 0;

  while(1){
    if (!s.empty() && arr[status] == s.top()){
      s.pop();
      result[resultIndex] = '-';
      resultIndex++;
      status++;
    }
    else if (i<=n){
      s.push(i);
      result[resultIndex] = '+';
      resultIndex++;
      i++;
    }
    else{
      break;
    }
  }

  if (status > n){
    for (int j=0;j<resultIndex;j++){
      printf("%c\n",result[j]);
    }
  }
  else{
    printf("NO");
  }

  return 0;
}
